docs/sources/developers/plugins/add-support-for-variables.md

```diff
@@ -58,3 +58,97 @@ For data sources, you need to use the [getTemplateSrv]({{< relref "../../package
+
+Let's start by defining a query model for the variable query.
+
+```ts
+export interface MyVariableQuery {
+  namespace: string;
+  rawQuery: string;
+}
+```
+
+For a data source to support query variables you need to override the [`metricFindQuery`]({{< relref "../../packages_api/data/datasourceapi.md#metricfindquery-method" >}}) in your `DataSourceApi` class. `metricFindQuery` returns an array of [`MetricFindValue`]({{< relref "../../packages_api/data/metricfindvalue.md" >}}) which has a single property, `text`.
+
+```ts
+async metricFindQuery(query: MyVariableQuery, options?: any) {
+  // Retrieve DataQueryResponse based on query.
+  const response = await this.fetchMetricNames(query.namespace, query.rawQuery);
+
+  // Convert query results to a MetricFindValue[]
+  const values = response.data.map(frame => ({ text: frame.name }));
+
+  return values;
+}
+```
+
+> **Note:** By default, Grafana provides a default query model and editor for simple text queries. If that's all you need, you can leave the query type as `string`.
+>
+> ```ts
+> async metricFindQuery(query: string, options?: any)
+> ```
+
+Let's create a custom query editor to allow the user to edit the query model.
+
+1. Create a `VariableQueryEditor` component.
+
+   ```ts
+   import React, { useState } from 'react';
+   import { MyVariableQuery } from './types';
+
+   interface VariableQueryProps {
+     query: MyVariableQuery;
+     onChange: (query: MyVariableQuery, definition: string) => void;
+   }
+
+   export const VariableQueryEditor: React.FC<VariableQueryProps> = ({ onChange, query }) => {
+     const [state, setState] = useState(query);
+
+     const saveQuery = () => {
+       onChange(state, `${state.query} (${state.namespace})`);
+     };
+
+     const handleChange = (event: React.FormEvent<HTMLInputElement>) =>
+       setState({
+         ...state,
+         [event.currentTarget.name]: event.currentTarget.value,
+       });
+
+     return (
+       <>
+         <div className="gf-form">
+           <span className="gf-form-label width-10">Namespace</span>
+           <input
+             name="namespace"
+             className="gf-form-input"
+             onBlur={saveQuery}
+             onChange={handleChange}
+             value={state.namespace}
+           />
+         </div>
+         <div className="gf-form">
+           <span className="gf-form-label width-10">Query</span>
+           <input name="rawQuery" className="gf-form-input" onBlur={saveQuery} onChange={handleChange} value={state.rawQuery} />
+         </div>
+       </>
+     );
+   };
+   ```
+
+   Grafana saves the query model whenever one of the text fields loses focus (`onBlur`) and then previews the values returned by `metricFindQuery`.
+
+   The second argument to `onChange` allows you to set a text representation of the query which will appear next to the name of the variable in the variables list.
```

Comment:
I recommend putting the explanation for what the code block does above the code block or as code comments in the code block.

Reply:
I put them before initially, but since they are more notes/comments, I felt they wasn't valuable until the reader has read the code block.

Reply:
A good rule is to explain to the reader what they are reading before they read it, not after.